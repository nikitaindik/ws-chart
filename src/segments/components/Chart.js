import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';

import {
  VictoryChart,
  VictoryBar,
  VictoryLine,
  VictoryAxis,
  VictoryScatter,
  VictoryVoronoiContainer,
} from 'victory';

import { timeFormat } from 'd3-time-format';

import { receiveHistory, receiveUpdate, changeMode } from '../actions';
import { socketEventTypes } from '../constants';

import { selectDisplayData } from '../selectors';

import socketModule from '../../setupSocket';

import Button from '../../components/Button';

import TimeIcon from './TimeIcon';

const oneDayInMs = 1 * 24 * 60 * 60 * 1000;
const sixHoursInMs = 6 * 60 * 60 * 1000;
const oneHourInMs = 1 * 60 * 60 * 1000;

const mapStateToProps = state => ({
  availableData: selectDisplayData(state),
  activeBarSize: state.segments.activeBarSize,
});

const mapDispatchToProps = dispatch => ({
  changeMode: mode => dispatch(changeMode(mode)),
  dispatch,
});

const LegendIconSquare = ({ color }) => (
  <div style={{ background: color, width: '16px', height: '16px' }} />
);

const LegendIconLine = ({ color }) => (
  <div
    style={{
      width: '16px',
      height: '16px',
      display: 'flex',
      alignItems: 'center',
    }}
  >
    <div style={{ background: color, height: '3px', width: '100%' }} />
  </div>
);

const LegendItem = ({ icon, label, value, primary }) => (
  <div style={{ display: 'flex', marginBottom: '8px' }}>
    <div style={{ marginRight: '8px' }}>{icon}</div>
    <div style={{ marginRight: '8px' }}>{label}:</div>
    <div style={{ fontWeight: primary ? 'bold' : 'normal' }}>{value}</div>
  </div>
);

const Legend = ({ activeBarData }) => {
  if (!activeBarData) {
    return <div style={{ height: '100px' }}>Hover over chart</div>;
  }

  const { added, removed, segmentSize, timestamp } = activeBarData;

  const formatter = timeFormat('%b %d, %H:%M');
  const formattedTime = formatter(timestamp);

  const columnStyle = {
    padding: '20px',
    width: '200px',
  };

  return (
    <div
      style={{
        height: '100px',
        display: 'flex',
        textAlign: 'left',
        justifyContent: 'center',
      }}
    >
      <div style={columnStyle}>
        <LegendItem
          icon={<LegendIconSquare color="#6c59d3" />}
          label={'Added'}
          value={added}
        />
        <LegendItem
          icon={<LegendIconSquare color="#999ea1" />}
          label={'Removed'}
          value={Math.abs(removed)}
        />
      </div>
      <div style={columnStyle}>
        <LegendItem
          icon={<LegendIconLine color="#008af7" />}
          label={'Size'}
          value={segmentSize}
          primary
        />
        <LegendItem icon={<TimeIcon />} label={'Time'} value={formattedTime} />
      </div>
    </div>
  );
};

const ChartContainer = ({
  activeBarSize,
  availableData,
  dispatch,
  changeMode,
}) => {
  useEffect(() => {
    socketModule.subscribe(event => {
      switch (event.type) {
        case socketEventTypes.SEGMENT_HISTORY:
          const { segmentId, data } = event.payload;
          dispatch(receiveHistory(segmentId, data));
          break;
        case socketEventTypes.SEGMENT_UPDATE:
          dispatch(receiveUpdate(event.data));
          break;
        default:
      }
    });
  }, []);

  const [activeBarData, setActiveBarData] = useState(null);

  if (!availableData) {
    return null;
  }

  return (
    <>
      <Legend activeBarData={activeBarData} />
      <div style={{ height: '400px' }}>
        <Chart data={availableData} setActiveBarData={setActiveBarData} />
      </div>
      <ChartModeControls
        activeBarSize={activeBarSize}
        onChangeModeClick={changeMode}
      />
    </>
  );
};

const ConnectedChartContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ChartContainer);

export default ConnectedChartContainer;

const ChartModeControls = ({ activeBarSize, onChangeModeClick }) => {
  return (
    <div style={{ display: 'flex', justifyContent: 'center' }}>
      <Button
        isActive={activeBarSize === 86400000}
        onClick={() => onChangeModeClick(oneDayInMs)}
      >
        Last 30 days
      </Button>
      <Button
        isActive={activeBarSize === 21600000}
        onClick={() => onChangeModeClick(sixHoursInMs)}
      >
        Last 7 days
      </Button>
      <Button
        isActive={activeBarSize === 3600000}
        onClick={() => onChangeModeClick(oneHourInMs)}
      >
        Last 24 hours
      </Button>
      <Button
        isActive={activeBarSize === 300000}
        onClick={() => onChangeModeClick(300000)}
      >
        Last 60 minutes
      </Button>
    </div>
  );
};

const Chart = ({ data, setActiveBarData }) => (
  <VictoryChart
    padding={{ top: 0, left: 40, right: 40, bottom: 50 }}
    domainPadding={20}
    scale={{ x: 'time' }}
    height={300}
    width={700}
    containerComponent={
      <VictoryVoronoiContainer
        voronoiDimension="x"
        onActivated={points => {
          const { added, removed, segmentSize, timestamp } = points[0];
          setActiveBarData({
            added,
            removed,
            segmentSize,
            timestamp,
          });
        }}
      />
    }
  >
    <VictoryAxis offsetY={50} />
    <VictoryAxis
      animate
      dependentAxis
      style={{
        grid: {
          stroke: '#999ea1',
          opacity: 0.5,
        },
      }}
    />
    <VictoryBar
      data={data}
      x="timestamp"
      y="added"
      name="added"
      style={{ data: { fill: '#6c59d3' } }}
      animate
    />
    <VictoryBar
      data={data}
      x="timestamp"
      y="removed"
      name="removed"
      style={{ data: { fill: '#999ea1' } }}
      animate
    />
    <VictoryLine
      data={data}
      x="timestamp"
      y="segmentSize"
      name="segmentSize"
      interpolation="natural"
      style={{ data: { stroke: '#008af7' } }}
      // animate
    />
    <VictoryScatter
      data={data}
      x="timestamp"
      y="segmentSize"
      size={(datum, active) => (active ? 3 : 0)}
      style={{ data: { fill: '#008af7', stroke: '#fff', strokeWidth: '1px' } }}
      // animate
    />
  </VictoryChart>
);
