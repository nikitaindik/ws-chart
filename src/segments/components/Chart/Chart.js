import React from 'react';

import {
  VictoryChart,
  VictoryBar,
  VictoryLine,
  VictoryAxis,
  VictoryScatter,
  VictoryVoronoiContainer,
} from 'victory';

import { colors } from '../../constants';

import style from './Chart.module.css';

const Chart = ({ data, setActiveBarData, setIsHovered }) => (
  <div
    className={style.chartWrap}
    onMouseEnter={() => setIsHovered(true)}
    onMouseLeave={() => setIsHovered(false)}
  >
    <VictoryChart
      animate={false}
      padding={{ top: 20, left: 40, right: 40, bottom: 50 }}
      domainPadding={20}
      scale={{ x: 'time' }}
      height={300}
      width={700}
      containerComponent={
        <VictoryVoronoiContainer
          voronoiDimension="x"
          onActivated={points => {
            const { added, removed, segmentSize, timestamp } = points[0];
            setActiveBarData({
              added,
              removed,
              segmentSize,
              timestamp,
            });
          }}
        />
      }
    >
      <VictoryAxis offsetY={50} />
      <VictoryAxis
        // animate
        dependentAxis
        style={{
          grid: {
            stroke: '#999ea1',
            opacity: 0.5,
          },
        }}
      />
      <VictoryBar
        data={data}
        x="timestamp"
        y="added"
        name="added"
        style={{ data: { fill: colors.CHART_POSITIVE_COLOR } }}
        // animate
      />
      <VictoryBar
        data={data}
        x="timestamp"
        y="removed"
        name="removed"
        style={{ data: { fill: colors.CHART_NEGATIVE_COLOR } }}
        // animate
      />
      <VictoryLine
        data={data}
        x="timestamp"
        y="segmentSize"
        name="segmentSize"
        interpolation="natural"
        style={{ data: { stroke: colors.CHART_LINE_COLOR } }}
        // animate
      />
      <VictoryScatter
        data={data}
        x="timestamp"
        y="segmentSize"
        size={(datum, active) => (active ? 5 : 0)}
        style={{
          data: {
            fill: colors.CHART_LINE_COLOR,
            stroke: '#fff',
            strokeWidth: '1px',
          },
        }}
        // animate
      />
    </VictoryChart>
  </div>
);

export default Chart;
