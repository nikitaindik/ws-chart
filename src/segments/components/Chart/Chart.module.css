.chartWrap {
  height: 400px;
}
