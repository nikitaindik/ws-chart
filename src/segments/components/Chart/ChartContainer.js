import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';

import { receiveHistory, receiveUpdate } from '../../actions';
import { socketEventTypes } from '../../constants';

import { selectDisplayData, selectActiveBarSize } from '../../selectors';

import webSocketConnection from '../../../core/webSocketConnection';

import ChartLegend from '../ChartLegend/ChartLegend';
import Chart from './Chart';
import ChartModeContainer from '../ChartMode/ChartModeContainer';

import mockWebSocketServer from '../../../core/mockWebSocketServer';
import mockServerResponses from '../../utils/mockServerResponses';

if (process.env.REACT_APP_USE_MOCK_SERVER === 'true') {
  // Use mock server for development, real server - for production
  mockWebSocketServer.subscribe(mockServerResponses);
}

const handleSocketEvents = dispatch => {
  return webSocketConnection.subscribe(event => {
    switch (event.type) {
      case socketEventTypes.SEGMENT_HISTORY:
        const { segmentId, data } = event.payload;
        dispatch(receiveHistory(segmentId, data));
        break;
      case socketEventTypes.SEGMENT_UPDATE:
        dispatch(receiveUpdate(event.data));
        break;
      default:
    }
  });
};

const ChartContainer = ({ dataToDisplay, activeBarSize, dispatch }) => {
  useEffect(() => handleSocketEvents(dispatch), []);

  const [hoveredBarData, setHoveredBarData] = useState(null);
  const [isHovered, setIsHovered] = useState(false);

  if (!dataToDisplay) {
    return null;
  }

  const latestBarData = dataToDisplay[dataToDisplay.length - 1];
  const isHoveringOverLastBar =
    hoveredBarData && hoveredBarData.timestamp === latestBarData.timestamp;

  return (
    <>
      <ChartLegend
        hoveredBarData={hoveredBarData}
        showLatestBarData={!isHovered || isHoveringOverLastBar}
        latestBarData={latestBarData}
        isHoveringOverLastBar={isHoveringOverLastBar}
        activeBarSize={activeBarSize}
      />
      <Chart
        data={dataToDisplay}
        setActiveBarData={setHoveredBarData}
        setIsHovered={setIsHovered}
      />
      <ChartModeContainer />
    </>
  );
};

const mapStateToProps = state => ({
  dataToDisplay: selectDisplayData(state),
  activeBarSize: selectActiveBarSize(state),
});

export default connect(mapStateToProps)(ChartContainer);
