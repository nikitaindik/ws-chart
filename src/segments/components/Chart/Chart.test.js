import React from 'react';
import { render } from 'react-testing-library';

import Chart from './Chart';

test('Chart snapshot', () => {
  const mockedProps = {
    data: [
      { timestamp: 1, segmentId: '10', added: 1, removed: 1, segmentSize: 1 },
    ],
    setActiveBarData: jest.fn(),
    setIsHovered: jest.fn(),
  };

  const { container } = render(<Chart {...mockedProps} />);
  expect(container).toMatchSnapshot();
});
