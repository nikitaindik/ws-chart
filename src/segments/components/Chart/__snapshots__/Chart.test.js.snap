// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart snapshot 1`] = `
<div>
  <div
    class="chartWrap"
  >
    <div
      class="VictoryContainer"
      style="height: 100%; width: 100%; user-select: none; pointer-events: none; position: relative;"
    >
      <svg
        aria-labelledby="victory-container-1-title victory-container-1-desc"
        height="300"
        role="img"
        style="pointer-events: all; width: 100%; height: 100%;"
        viewBox="0 0 800 300"
        width="800"
      >
        <g
          role="presentation"
        >
          <line
            role="presentation"
            shape-rendering="auto"
            style="stroke: #252525; fill: transparent; stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
            vector-effect="non-scaling-stroke"
            x1="40"
            x2="760"
            y1="250"
            y2="250"
          />
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: none; fill: none; pointer-events: painted;"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="40"
              y1="250"
              y2="20"
            />
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: transparent; fill: transparent; size: 1px;"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="40"
              y1="250"
              y2="251"
            />
            <text
              direction="inherit"
              dx="0"
              dy="11.969999999999999"
              x="40"
              y="261"
            >
              <tspan
                dx="0"
                style="font-family: 'Gill Sans', 'Gill Sans MT', 'Seravek', 'Trebuchet MS', sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; fill: #252525; stroke: transparent;"
                text-anchor="middle"
                x="40"
              >
                02 AM
              </tspan>
            </text>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: none; fill: none; pointer-events: painted;"
              vector-effect="non-scaling-stroke"
              x1="760"
              x2="760"
              y1="250"
              y2="20"
            />
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: transparent; fill: transparent; size: 1px;"
              vector-effect="non-scaling-stroke"
              x1="760"
              x2="760"
              y1="250"
              y2="251"
            />
            <text
              direction="inherit"
              dx="0"
              dy="11.969999999999999"
              x="760"
              y="261"
            >
              <tspan
                dx="0"
                style="font-family: 'Gill Sans', 'Gill Sans MT', 'Seravek', 'Trebuchet MS', sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; fill: #252525; stroke: transparent;"
                text-anchor="middle"
                x="760"
              >
                .001
              </tspan>
            </text>
          </g>
        </g>
        <g
          role="presentation"
        >
          <line
            role="presentation"
            shape-rendering="auto"
            style="stroke: #252525; fill: transparent; stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
            vector-effect="non-scaling-stroke"
            x1="40"
            x2="40"
            y1="20"
            y2="250"
          />
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: #999ea1; opacity: 0.5; fill: none; pointer-events: painted;"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="760"
              y1="207.9723661527347"
              y2="207.9723661527347"
            />
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: transparent; fill: transparent; size: 1px;"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="39"
              y1="207.9723661527347"
              y2="207.9723661527347"
            />
            <text
              direction="inherit"
              dx="0"
              dy="4.97"
              x="29"
              y="207.9723661527347"
            >
              <tspan
                dx="0"
                style="font-family: 'Gill Sans', 'Gill Sans MT', 'Seravek', 'Trebuchet MS', sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; fill: #252525; stroke: transparent;"
                text-anchor="end"
                x="29"
              >
                0.2
              </tspan>
            </text>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: #999ea1; opacity: 0.5; fill: none; pointer-events: painted;"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="760"
              y1="165.94473230546942"
              y2="165.94473230546942"
            />
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: transparent; fill: transparent; size: 1px;"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="39"
              y1="165.94473230546942"
              y2="165.94473230546942"
            />
            <text
              direction="inherit"
              dx="0"
              dy="4.97"
              x="29"
              y="165.94473230546942"
            >
              <tspan
                dx="0"
                style="font-family: 'Gill Sans', 'Gill Sans MT', 'Seravek', 'Trebuchet MS', sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; fill: #252525; stroke: transparent;"
                text-anchor="end"
                x="29"
              >
                0.4
              </tspan>
            </text>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: #999ea1; opacity: 0.5; fill: none; pointer-events: painted;"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="760"
              y1="123.91709845820415"
              y2="123.91709845820415"
            />
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: transparent; fill: transparent; size: 1px;"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="39"
              y1="123.91709845820415"
              y2="123.91709845820415"
            />
            <text
              direction="inherit"
              dx="0"
              dy="4.97"
              x="29"
              y="123.91709845820415"
            >
              <tspan
                dx="0"
                style="font-family: 'Gill Sans', 'Gill Sans MT', 'Seravek', 'Trebuchet MS', sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; fill: #252525; stroke: transparent;"
                text-anchor="end"
                x="29"
              >
                0.6
              </tspan>
            </text>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: #999ea1; opacity: 0.5; fill: none; pointer-events: painted;"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="760"
              y1="81.88946461093886"
              y2="81.88946461093886"
            />
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: transparent; fill: transparent; size: 1px;"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="39"
              y1="81.88946461093886"
              y2="81.88946461093886"
            />
            <text
              direction="inherit"
              dx="0"
              dy="4.97"
              x="29"
              y="81.88946461093886"
            >
              <tspan
                dx="0"
                style="font-family: 'Gill Sans', 'Gill Sans MT', 'Seravek', 'Trebuchet MS', sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; fill: #252525; stroke: transparent;"
                text-anchor="end"
                x="29"
              >
                0.8
              </tspan>
            </text>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: #999ea1; opacity: 0.5; fill: none; pointer-events: painted;"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="760"
              y1="39.86183076367357"
              y2="39.86183076367357"
            />
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: transparent; fill: transparent; size: 1px;"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="39"
              y1="39.86183076367357"
              y2="39.86183076367357"
            />
            <text
              direction="inherit"
              dx="0"
              dy="4.97"
              x="29"
              y="39.86183076367357"
            >
              <tspan
                dx="0"
                style="font-family: 'Gill Sans', 'Gill Sans MT', 'Seravek', 'Trebuchet MS', sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; fill: #252525; stroke: transparent;"
                text-anchor="end"
                x="29"
              >
                1.0
              </tspan>
            </text>
          </g>
        </g>
        <g
          role="presentation"
        >
          <path
            d="M 758, 250 
A 0 0 0 0 1, 758, 250 
L 758, 39.86183076367357 
A 0 0 0 0 1, 758, 39.86183076367357 
L 762, 39.86183076367357 
A 0 0 0 0 1, 762, 39.86183076367357 
L 762, 250 
A 0 0 0 0 1, 762, 250 
 z"
            role="presentation"
            shape-rendering="auto"
            style="fill: #6c59d3; stroke: #6c59d3; padding: 8px; stroke-width: 0;"
          />
        </g>
        <g
          role="presentation"
        >
          <path
            d="M 758, 250 
A 0 0 0 0 1, 758, 250 
L 758, 39.86183076367357 
A 0 0 0 0 1, 758, 39.86183076367357 
L 762, 39.86183076367357 
A 0 0 0 0 1, 762, 39.86183076367357 
L 762, 250 
A 0 0 0 0 1, 762, 250 
 z"
            role="presentation"
            shape-rendering="auto"
            style="fill: #999ea1; stroke: #999ea1; padding: 8px; stroke-width: 0;"
          />
        </g>
        <g
          clip-path="url(#victory-clip-2)"
          style="height: 100%; width: 100%; user-select: none;"
        >
          <defs>
            <clippath
              id="victory-clip-2"
            >
              <rect
                height="230"
                role="presentation"
                shape-rendering="auto"
                vector-effect="non-scaling-stroke"
                width="720"
                x="40"
                y="20"
              />
            </clippath>
          </defs>
          <path
            role="presentation"
            shape-rendering="auto"
            style="fill: transparent; stroke: #008af7; stroke-width: 2;"
          />
        </g>
        <g>
          <path
            d="M 760, 39.86183076367357
      m 0, 0
      a 0, 0 0 1,0 0,0
      a 0, 0 0 1,0 0,0"
            role="presentation"
            shape-rendering="auto"
            style="fill: #008af7; stroke: #fff; stroke-width: 1px;"
          />
        </g>
      </svg>
      <div
        style="z-index: 99; position: absolute; top: 0px; left: 0px; width: 100%; height: 100%;"
      >
        <svg
          height="300"
          style="overflow: visible; width: 100%; height: 100%;"
          viewBox="0 0 800 300"
          width="800"
        />
      </div>
    </div>
  </div>
</div>
`;
